tests/test_connection.py

```diff
@@ -848,7 +849,37 @@ def test_funding_external_wallet_corners(node_factory, bitcoind):
 
     l1.rpc.fundchannel_cancel(l2.info['id'])
     # Should be able to 'restart' after canceling
-    l1.rpc.fundchannel_start(l2.info['id'], amount)
+    amount2 = 1000000
+    funding_addr = l1.rpc.fundchannel_start(l2.info['id'], amount2)['funding_address']
+
+    addr = l1.rpc.newaddr()['bech32']
+    l1.bitcoin.rpc.sendtoaddress(addr, 0.1)
+    bitcoind.generate_block(1)
+    wait_for(lambda: len(l1.rpc.listfunds()['outputs']) == 1)
+    # Create the funding transaction
+    prep = l1.rpc.txprepare(funding_addr, amount2)
+    decode = bitcoind.rpc.decoderawtransaction(prep['unsigned_tx'])
+    assert decode['txid'] == prep['txid']
+
+    # One output will be correct.
+    if decode['vout'][0]['value'] == Decimal('0.01000000'):
+        txout = 0
+    elif decode['vout'][1]['value'] == Decimal('0.01000000'):
+        txout = 1
+    else:
+        assert False
+
+    # Be sure fundchannel_complete is successful
+    assert l1.rpc.fundchannel_complete(l2.info['id'], prep['txid'], txout)['commitments_secured']
+    # Canceld channel after fundchannel_complete
+    assert l1.rpc.fundchannel_cancel(l2.info['id'])['cancelled']
+
+    l1.rpc.connect(l2.info['id'], 'localhost', l2.port)
+    l1.rpc.fundchannel_start(l2.info['id'], amount)['funding_address']
+    assert l1.rpc.fundchannel_complete(l2.info['id'], prep['txid'], txout)['commitments_secured']
+    l1.rpc.txsend(prep['txid'])
```

Comment:
should we also check that it catches that the tx was broadcast by a third party, like bitcoind?

Reply:
This is a belt and suspenders check. I  theory the caller MUST NOT call `fundchannel_cancel` after any possibility of broadcast.

Reply:
@ZmnSCPxj  Maybe the transaction is blocked in the txmempool of the miner node?

For now, we can't fully ensure that the broadcast was catched from a third pary. Only when the transaction (broadcast by a third pary) is onchain, we can catch it.

On the other hand, because of f77d6c7 , the transaction broadcast by `lightningd` will have a entry in DB so we can catch it.

Reply:
in lightningd/channel_control.c (line 669-677):
```
	enum wallet_tx_type type;
	if(wallet_transaction_type(cmd->ld->wallet,
				   &cancel_channel->funding_txid,
				   &type))
		return command_fail(cmd, LIGHTNINGD,
				    "Has the funding transaction been broadcast? "
				    "Please use `close` or `dev-fail` instead.");
```
Here we try to catch the tx broadcast by `lightningd`. We can make sure if the tx is broadcast by `lightningd`.

and in line690-698:
```
	/* Check if the transaction is onchain. */
	/* Note: The above check and this check can't completely ensure that
	 * the funding transaction isn't broadcast. We can't know if the funding
	 * is broadcast by external wallet and the transaction hasn't been onchain. */
	bitcoind_gettxout(cmd->ld->topology->bitcoind,
			  &cancel_channel->funding_txid,
			  cancel_channel->funding_outnum,
			  process_check_funding_broadcast,
			  cancel_channel);
```
Here we check the tx broadcast by a third party. But it's a `incomplete` check because we can't catch the case if the tx hasn't been onchain.

Reply:
> and in line690-698:

Yeah, I saw that. it's not foolproof, but not a bad check. i was suggesting that if we're going to include the bitcoind check we should include a test to verify it works as intended.

Reply:
Ok! I'll add the check for bitcoind ;-)